### DISCUSSION TITLE
Added IMU to my diffbot

### DISCUSSION THREAD

**Discussion by joeuser846** [fjp/2wd-robot#59] [Ideas]
Thought you might be interested that I successfully added an IMU to my diffbot using the Adafruit [BNO055](https://www.adafruit.com/product/2472?gclid=EAIaIQobChMIqITG8cXG9gIVlRXUAR38Cgi5EAAYASAAEgL65PD_BwE) breakout board.  This board supports I2C to the Rpi or Jetson Nano on the diffbot and works with [this](https://github.com/dheera/ros-imu-bno055) driver.  Just be sure to follow the instructions [here](https://learn.adafruit.com/circuitpython-on-raspberrypi-linux/i2c-clock-stretching) if using a Rpi.

I then use robot_localization to fuse the odom from ros_control as implemented in diffbot with the IMU.  I basically tell robot_localization to ignore yaw input from the motor encoders and substitute yaw from the IMU data.  It beautifully fixes the really bad odom from my tank/track based diffbot variant.  Tracked vehicles have a major problem with slipping when turning that makes pure encoder odom pretty useless.

I was hung-up for a while trying to get the new transform tree sorted until I discovered a parameter in diffbot_control .yaml that allows you to suppress publishing of the base_footprint -> odom transform which then permits robot_localization do it instead without any conflicts.  This fixed my final problem:

    # Top level frame (link) of the robot description
    # Don't publish base_footprint->odom transform here. It will be done by robot_localization after sensor fusion.
    # base_frame_id: base_footprint 
    enable_odom_tf: false

I can now drive and navigate around with very good accuracy despite the inevitable track slippage.

**Comment by fjp** [comment]
Thank you @joeuser846 for sharing, this is very useful! I will hopefully soon integrate the IMU as well. Very interesting to know that this "high level" approach (connecting the IMU directly to the SBC) works well for you! 

An alternative way I planned to do it is to connect the IMU to the micro controller (Teensy in my case), use a suitable low-level library and finish implementing the main.cpp script of the `base_controller`:

https://github.com/ros-mobile-robots/diffbot/blob/2f154a8967421a9698a9b0a1d8e3ae718af607f1/diffbot_base/scripts/base_controller/src/main.cpp#L53-L71

However, I might as well start out with the high level method.

Could you also please share some details on how much effort it was to tune `robot_localization` correctly or what you had to take care of?

**Reply by fjp** [reply]
Reference to still open IMU issue [#21](https://github.com/ros-mobile-robots/diffbot/issues/21#issuecomment-751623420).

**Reply by joeuser846** [reply]
I just uploaded a pull request containing my config files.  Hopefully, you can get the gist of what I did by reviewing those.  I certainly don't claim my solution is optimized.  There are hundreds of parameters in this system that could be tweaked endlessly.  I had to fix a lot of problems to get this working so when it started behaving well I stopped and decided to enjoy the results for a while!
